components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -896,222 +895,65 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
-             epsi = ((f1pr05+f1pr14*sc**thrd*(rhofaci(i,k)*rho(i,k)/mu)**0.5_rtype)*2._rtype*pi* &
-                  rho(i,k)*dv)*nitot_incld(i,k)
-             epsi_tot   = epsi_tot + epsi
-          else
-             epsi = 0._rtype
-          endif
-
+          call ice_relaxation_timescale(rho(i,k),t(i,k),rhofaci(i,k),&
+          f1pr05,f1pr14,dv,mu,sc,qitot_incld(i,k),nitot_incld(i,k),&
+               epsi,epsi_tot)
 
           !.........................
           ! calculate rime density
-
-          !     FUTURE:  Add source term for birim (=qccol/rhorime_c) so that all process rates calculations
-          !              are done together, before conservation.
-
-          ! NOTE: Tc (ambient) is assumed for the surface temperature.  Technically,
-          ! we should diagose graupel surface temperature from heat balance equation.
-          ! (but the ambient temperature is a reasonable approximation; tests show
-          ! very little sensitivity to different assumed values, Milbrandt and Morrison 2012).
-
-          ! Compute rime density: (based on parameterization of Cober and List, 1993 [JAS])
-          ! for simplicty use mass-weighted ice and droplet/rain fallspeeds
-
-          ! if (qitot_incld(i,k).ge.qsmall .and. t(i,k).lt.zerodegc) then
-          !  NOTE:  condition applicable for cloud only; modify when rain is added back
-          if (qccol.ge.qsmall .and. t(i,k).lt.zerodegc) then
-
-             ! get mass-weighted mean ice fallspeed
-             vtrmi1(i,k) = f1pr02*rhofaci(i,k)
-             iTc   = 1._rtype/min(-0.001_rtype,t(i,k)-zerodegc)
-
-             ! cloud:
-             if (qc_incld(i,k).ge.qsmall) then
-                ! droplet fall speed
-                ! (use Stokes' formulation (thus use analytic solution)
-                Vt_qc(i,k) = acn(i,k)*gamma(4._rtype+bcn+mu_c(i,k))/(lamc(i,k)**bcn*gamma(mu_c(i,k)+4._rtype))
-                ! use mass-weighted mean size
-                D_c = (mu_c(i,k)+4._rtype)/lamc(i,k)
-                V_impact  = abs(vtrmi1(i,k)-Vt_qc(i,k))
-                Ri        = -(0.5e+6_rtype*D_c)*V_impact*iTc
-                !               Ri        = max(1.,min(Ri,8.))
-                Ri        = max(1.,min(Ri,12._rtype))
-                if (Ri.le.8.) then
-                   rhorime_c  = (0.051_rtype + 0.114_rtype*Ri - 0.0055_rtype*Ri**2)*1000._rtype
-                else
-                   ! for Ri > 8 assume a linear fit between 8 and 12,
-                   ! rhorime = 900 kg m-3 at Ri = 12
-                   ! this is somewhat ad-hoc but allows a smoother transition
-                   ! in rime density up to wet growth
-                   rhorime_c  = 611._rtype+72.25_rtype*(Ri-8._rtype)
-                endif
-
-             endif    !if qc>qsmall
-
-          else
-             rhorime_c = 400._rtype
-          endif ! qi > qsmall and T < 273.15
+          call rime_density(t(i,k),rhofaci(i,k),&
```

Comment:
I was going to complain that this should be prepended with the variables it affects, but then realized that rime density/fraction isn't a prognostic variable. Should we prepend 'helper functions' like this with "calc_" to make clear that they are performing intermediate calculations rather than representing a process?

Reply:
I think this is a great suggestion and will update the code and PR accordingly.